METRON-1489: Retrofit UI tests to run reliably

sardell opened the pull request:
## Contributor Comments
This Pull Request includes work started by @iraghumitra on his branch `e2e-using-await-1`. In addition to the asyn/await work he did, I added [protractor-flake](https://github.com/NickTomlin/protractor-flake) in order to re-run failing tests to ensure they are actual failures and not the result of flakey tests.

In order to test locally, be sure to install protractor-flake before running the e2e tests. This is done by running `npm install` inside of the metron-alerts directory.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

simonellistonball commented on metron-interface/metron-alerts/e2e/alert-details/alert-status/alert-details-status.e2e-spec.ts:
```diff
@@ -23,64 +23,67 @@ import {loadTestData, deleteTestData} from '../../utils/e2e_util';
-    expect(page.getAlertStatus('ESCALATE')).toEqual('RESOLVE');
-    expect(listPage.getAlertStatusById(alertId)).toEqual('RESOLVE');
-    page.clickNew();
+  xit('should change alert statuses', async function() : Promise<any> {
```
Any particular reason to skip this test?

sardell replied:
This change is from @iraghumitra's original work, so it's probably best he answers this question.

simonellistonball commented on metron-interface/metron-alerts/e2e/alert-details/alert-status/alert-details-status.e2e-spec.ts:
```diff
@@ -101,7 +104,7 @@ describe('metron-alerts alert status', function() {
   });
 
-  it('should add comments for tree view', () => {
+  xit('should add comments for tree view', () => {
```
Any particular reason to skip this test?

sardell replied:
@iraghumitra Same question as above.

nickwallen commented on metron-interface/metron-alerts/README.md:
```diff
@@ -126,7 +126,14 @@ The application will be available at http://host:4201 assuming the port is set t
 ## E2E Tests
 
-An expressjs server is available for mocking the elastic search api.
+#Caveats
```
This markdown does not render properly.  I think you need a space and it probably needs to be 3 hashes as in "### Caveats".

nickwallen replied:
This is still a problem @sardell

mmiklavc commented:
@sardell This is great to see - we may get our e2e tests running reliably as part of our build yet! In your intro notes, you mention doing an npm install to get the protractor-flake package installed. I believe we have Maven commands that handle installing these packages during the e2e test build. I think it's fine to sort out how we'll run the e2e tests in context of the full Metron build in a follow-on PR, but for this PR I think we should get that manual step to work consistently with how we install the other packages.

Small tip on Jira and PR titles - the community is likely not going to care or know about private QE environments (regardless of them being appreciative of those external efforts :) ). It's probably better to describe this Jira/PR as getting it stable to run as part of a daily Travis build, or something along those lines.

nickwallen commented:
Maybe I am wrong, but I thought these tests used to run headless.  It looks to me that these are running "headful" now. :)  

Either way, can these tests run headless or is there a major hurdle stopping us from doing that?

anandsubbu commented on metron-interface/metron-alerts/e2e/utils/clean_metron_update_table.ts:
```diff
@@ -0,0 +1,52 @@
+  return  new Promise(
+      function (resolve, reject) {
+        // resolve();
+        cleanupTable(resolve, reject);
```
IMHO, it would be better to default to `resolve()` only owing to the issues with Vagrant `ssh2` client.

anandsubbu commented:
@nickwallen , in order to get past the following error you are seeing..
```
<snip>
[13:44:09] E/launcher - All configured authentication methods failed
[13:44:09] E/launcher - Error: All configured authentication methods failed
    at tryNextAuth (/Users/nallen/tmp/metron-pr1004/metron-interface/metron-alerts/node_modules/ssh2/lib/client.js:380:17)
    at SSH2Stream.onUSERAUTH_FAILURE (/Users/nallen/tmp/metron-pr1004/metron-interface/metron-alerts/node_modules/ssh2/lib/client.js:591:5)
    at SSH2Stream.emit (events.js:160:13)
    at parsePacket (/Users/nallen/tmp/metron-pr1004/metron-interface/metron-alerts/node_modules/ssh2-streams/lib/ssh.js:3929:10)
    at SSH2Stream._transform (/Users/nallen/tmp/metron-pr1004/metron-interface/metron-alerts/node_modules/ssh2-streams/lib/ssh.js:669:13)
    at SSH2Stream.Transform._read (_stream_transform.js:185:10)
    at SSH2Stream._read (/Users/nallen/tmp/metron-pr1004/metron-interface/metron-alerts/node_modules/ssh2-streams/lib/ssh.js:251:15)
    at SSH2Stream.Transform._write (_stream_transform.js:173:12)
    at doWrite (_stream_writable.js:407:12)
    at writeOrBuffer (_stream_writable.js:393:5)
[13:44:09] E/launcher - Process exited with error code 199

```
The error you are seeing is because of the ssh2 module which authenticates to the vagrant node to do cleanup tasks. This needs to be fixed.

As a workaround, you can do the following in (metron-interface/metron-alerts/e2e/utils/clean_metron_update_table.ts) i.e. https://github.com/apache/metron/pull/1004/files#diff-2cc98b255ddf3a954ad22dfa2b5545fb

```
+        resolve();
+        //cleanupTable(resolve, reject);
```
i.e. you need to uncomment the `resolve()` function and comment out ` cleanupTable()`.

anandsubbu commented on metron-interface/metron-alerts/protractor.conf.js:
```diff
@@ -32,12 +33,13 @@ exports.config = {
   capabilities: {
     'browserName': 'chrome',
     'chromeOptions': {
+      args: ["--disable-gpu", "--window-size=1435,850" ],
```
`-headless` can be included here to default to headless execution.

sardell commented on metron-interface/metron-alerts/e2e/alerts-list/alert-status/alerts-list-status.e2e-spec.ts:
```diff
@@ -44,84 +45,84 @@ describe('metron-alerts alert status', function() {
+    await page.toggleAlertInList(7);
+    await page.toggleAlertInList(8);
+    await page.clickActionDropdownOption('Escalate');
+    // expect(await page.getAlertStatus(6, 'NEW')).toEqual('ESCALATE');
```
@iraghumitra what was the reason for commenting out this test?

sardell commented on metron-interface/metron-alerts/e2e/alerts-list/alerts-list.po.ts:
```diff
@@ -272,63 +308,149 @@ export class MetronAlertsPage {
+
+  async selectQuickTimeRangeAndGetTimeRangeAndTimeText(quickRange: string) {
+    let currentTimeRangeVal: any = [];
+    // return element.all(by.css('app-time-range button span')).getText()
```
@iraghumitra what was the reason for commenting out this test?

sardell commented:
@nickwallen This work is ready on my end. I'm happy with where I'm at since addressing the feedback given by others here. The automated retries are to be expected and means protractor-flake is doing its job.

nickwallen commented on metron-interface/metron-alerts/e2e/alerts-list/tree-view/tree-view.po.ts:
```diff
@@ -97,21 +155,24 @@ export class TreeViewPage {
-      return element.all(by.css('[data-name="' + name + '"] table tbody tr')).get(rowId).all(by.css('td')).getText();
-    });
-  }
+  // getDashGroupTableValuesForRow(name: string, rowId: number) {
```
There are still blocks of code commented out that need cleaned up.  Maybe you didn't push some additional commits out?

sardell replied:
@nickwallen No, this looks like something I need to address. Thanks for pointing this out. Previously, I only addressed the specific lines that were called out before in comments. I'll run through all currently commented tests in this PR, update and have @iraghumitra review since this stems from the work he did. I also see some console.log's in there that I'll remove.

nickwallen commented on metron-interface/metron-alerts/e2e/alerts-list/meta-alerts/meta-alert.e2e-spec.ts:
```diff
@@ -19,139 +19,142 @@
+    expect(await tablePage.getHiddenRowCount()).toEqualBcoz(0, '0 rows to be hidden after group expand');
+
+    /* Meta Alert Status Change // These do not work till patch works */
+    // await tablePage.toggleAlertInList(0);
```
What do these commented out sections mean?  What patch?  When will the patch work?

nickwallen commented:
Please fix the formatting issue with the README.  I have mentioned in another thread, but you might have missed it.

sardell commented:
Re-running Travis.

nickwallen commented:
@sardell Should the topologies be stopped in the development environment when running the tests?      Does that impact the tests at all?

nickwallen commented:
The first time I ran against a development environment where the topologies were running and so new data (outside the e2e tests) was hitting the search indices.  I ran the e2e tests and had quite a few failures.

Then I stopped the topologies and since then either all tests pass on the first try or a couple fail and the automated retry kicks in which leads to eventual success.

I am just not sure if that is coincidence or if we need to update the README so that everyone knows to stop the topologies when running these tests.

sardell commented:
@nickwallen I wasn't able to reproduce the failure you experienced, but I added a note to the README stating the parsers will sometimes cause failures for users.

nickwallen commented:
+1 Thanks for all the hard work on this one.  I know this was a battle to get it over the finish line.

I will get this merged once the CI build completes.
